xpk.py

```diff
@@ -2754,30 +2850,29 @@ def run_gke_node_pool_create_command(args, system) -> int:
       f'{node_pools_to_delete}. \nDo you wish to delete: y (yes) / n (no):\n')
 
   if not will_delete:
-    xpk_print('Skipping delete commands. Continuing to next step.')
-  else:
-    for existing_node_pool_name in node_pools_to_delete:
-      if (
-          existing_node_pool_name.find(f'{args.cluster}-np-') == 0
-          and existing_node_pool_name not in desired_node_pool_names
-      ):
-        command = (
-            'gcloud beta container node-pools delete'
-            f' {existing_node_pool_name} --cluster={args.cluster}'
-            f' --zone={zone_to_region(args.zone)}'
-            f' --project={args.project} --quiet'
-        )
-        task = f'Nodepool-Delete-{existing_node_pool_name}'
-        commands.append(command)
-        task_names.append(task)
+    xpk_print('You have requested to not delete the existing nodepools in the cluster.'
+              ' There will be no change to the cluster.')
+    xpk_exit(1)
 
-  for i, command in enumerate(commands):
-    xpk_print(f'To complete {task_names[i]} we are executing {command}')
+  # Deletion of nodepools should happen before attempting to create new nodepools for the case
+  # when cluster is getting updated from 'x' device_type to 'y' device_type. In that case,
+  # '{args.cluster}-np-i' nodepool will be re-created for 'y' device_type.
+  for i, command in enumerate(delete_commands):
+    xpk_print(f'To complete {delete_task_names[i]} we are executing {command}')
+  max_return_code = run_commands(
+      delete_commands, 'Delete Nodepools', delete_task_names, dry_run=args.dry_run
+  )
+  if max_return_code != 0:
+    xpk_print(f'Delete Nodepools returned ERROR {max_return_code}')
+    return 1
+
```

Comment:
I think we should update the config map after the deletion is successful with the new device_type.

There are a few scenarios that the node pool commands could succeed / fail, and this would help avoid unnecessary deletion of node pools.

1. User creates v4-8 cluster
2. User changes cluster to v5e-8

Outcome 1:
Command fails to delete, this is ok since the type of the cluster is still v4-8

Outcome 2:
Command succeeds to delete, but fails to create new. This is ok since there are 0 node pools so the type doesnt matter.

Outcome 3:
Command succeeds to delete and partially fails to create new. This is not ok since the tpu type doesnt align with what is in the clsuter. This can cause following calls to delete the node pools and retry creation from 0. We can avoid this by setting the device_type of the cluster after delete. 

What do you think?

Reply:
I am updating the ConfigMap after deletion is successful. 
So, the different scenarios look like this when user updated cluster from 2 slices of v4-8 to 2 slices of v5e-8:
ConfigMap before the update is called: `v4-8: 2`

*Scenario 1: Deletion of v4-8 is unsuccessful*
Outcome: ConfigMap will be `v4-8: 2`.

*Scenario 2: Deletion of v4-8 is successful but creation of v5e-8 fails*
Outcome: ConfigMap will be updated to `v5e-8: 0`.

*Scenario 3: Deletion of v4-8 is successful but creation of v5e-8 partially fails*
Outcome: ConfigMap will be updated to `v5e-8: 0`. The following call to `create cluster` will not delete the v5e-8 that were already created.